tools/seal_oem_partition.go

```diff
@@ -0,0 +1,204 @@
+	hashOffset := "--hash-offset=" + strconv.FormatUint(oemFSSize4K<<12, 10)
+	cmd := exec.Command("sudo", "docker", "run", "--rm", "--name", "veritysetup", "--privileged", "-v", "/dev:/dev", imageID, "veritysetup",
+		"format", "/dev/sda8", "/dev/sda8", "--data-block-size=4096", "--hash-block-size=4096", dataBlocks,
+		hashOffset, "--no-superblock", "--format=0")
+	var verityBuf bytes.Buffer
+	cmd.Stdout = &verityBuf
+	if err := cmd.Run(); err != nil {
+		return "", "", fmt.Errorf("error in running docker veritysetup, "+
+			"input: oemFSSize4K=%d, error msg: (%v)", oemFSSize4K, err)
+	}
+	// Output of veritysetup is like:
+	// VERITY header information for /dev/sdb1
+	// UUID:
+	// Hash type:              0
+	// Data blocks:            2048
+	// Data block size:        4096
+	// Hash block size:        4096
+	// Hash algorithm:         sha256
+	// Salt:                   9cd7ba29a1771b2097a7d72be8c13b29766d7617c3b924eb0cf23ff5071fee47
+	// Root hash:              d6b862d01e01e6417a1b5e7eb0eed2a2189594b74325dd0749cd83bbf78f5dc8
+	lines := strings.Split(verityBuf.String(), "\n")
+	if !strings.HasPrefix(lines[len(lines)-2], "Root hash:") || !strings.HasPrefix(lines[len(lines)-3], "Salt:") {
+		return "", "", fmt.Errorf("error in veritsetup output format, the last two lines are not \"Salt:\" and \"Root hash:\", "+
+			"input: oemFSSize4K=%d, veritysetup output: %s", oemFSSize4K, verityBuf.String())
+	}
+	hash := strings.TrimSpace(strings.Split(lines[len(lines)-2], ":")[1])
+	salt := strings.TrimSpace(strings.Split(lines[len(lines)-3], ":")[1])
+	return hash, salt, nil
+}
+
+// appendDMEntryToGRUB appends an dm-verity table entry to kernel command line in grub.cfg
+// A target line in grub.cfg looks like
+// ...... root=/dev/dm-0 dm="1 vroot none ro 1,0 4077568 verity payload=PARTUUID=8AC60384-1187-9E49-91CE-3ABD8DA295A7 hashtree=PARTUUID=8AC60384-1187-9E49-91CE-3ABD8DA295A7 hashstart=4077568 alg=sha256 root_hexdigest=xxxxxxxx salt=xxxxxxxx"
+func appendDMEntryToGRUB(grubPath, name, partUUID, hash, salt string, oemFSSize4K uint64) error {
+	grubPath = grubPath + "/grub.cfg"
+	// from 4K blocks to 512B sectors
+	oemFSSizeSector := oemFSSize4K << 3
+	entryString := fmt.Sprintf("%s none ro 1, 0 %d verity payload=PARTUUID=%s hashtree=PARTUUID=%s hashstart=%d alg=sha256 "+
+		"root_hexdigest=%s salt=%s\"", name, oemFSSizeSector, partUUID, partUUID, oemFSSizeSector, hash, salt)
+	grubContent, err := ioutil.ReadFile(grubPath)
+	if err != nil {
+		return fmt.Errorf("cannot read grub.cfg at %q, "+
+			"input: grubPath=%q, name=%q, partUUID=%q, oemFSSize4K=%d, hash=%q, salt=%q, "+
+			"error msg:(%v)", grubPath, grubPath, name, partUUID, oemFSSize4K, hash, salt, err)
+	}
+	lines := strings.Split(string(grubContent), "\n")
+	// add the entry to all kernel command lines containing "dm="
+	for idx, line := range lines {
+		if !strings.Contains(line, "dm=") {
+			continue
+		}
+		startPos := strings.Index(line, "dm=")
+		lineBuf := []rune(line[:len(line)-1])
```

Comment:
Does this remove a quote or a newline? Can you leave a comment saying which one?